lib/CL/devices/level0/pocl-level0.cc

```diff
@@ -0,0 +1,1227 @@
+        memcpy(Ptr, Binary, BinarySize);
+        program->program_il = Ptr;
+        program->program_il_size = BinarySize;
+      }
+
+      pocl_cache_create_program_cachedir(program, device_i, Binary, BinarySize,
+                                         ProgramBcPath);
+      convertProgramBcToSpv(ProgramBcPath, ProgramSpvPath);
+
+      if (!pocl_exists(ProgramSpvPath)) {
+        char ProgramSpvPathTemp[POCL_FILENAME_LENGTH];
+        pocl_cache_tempname(ProgramSpvPathTemp, ".spv", NULL);
+
+        pocl_write_file(ProgramSpvPathTemp, Binary, BinarySize, 0, 0);
+        POCL_RETURN_ERROR_ON(
+            !pocl_exists(ProgramSpvPathTemp), CL_BUILD_PROGRAM_FAILURE,
+            "failed to write SPIR-V file %s\n", ProgramSpvPathTemp);
+
+        pocl_rename(ProgramSpvPathTemp, ProgramSpvPath);
+      }
+    } else {
+#ifdef ENABLE_SPIR
+      if (bitcode_is_triple(Binary, BinarySize, "spir-unknown") ||
+          bitcode_is_triple(Binary, BinarySize, "spir64-unknown")) {
+        char ProgramSpvPathTemp[POCL_FILENAME_LENGTH];
+        char ProgramBcPathTemp[POCL_FILENAME_LENGTH];
+
+        pocl_cache_tempname(ProgramSpvPathTemp, ".spv", NULL);
+        pocl_cache_tempname(ProgramBcPathTemp, ".bc", NULL);
+        int err = pocl_write_file(ProgramBcPathTemp, Binary, BinarySize, 0, 0);
+        POCL_RETURN_ERROR_ON((err != 0), CL_BUILD_PROGRAM_FAILURE,
+                             "failed to write BC file into cache\n");
+
+        err = compileProgramBcToSpv(program, device_i, ProgramBcPathTemp,
+                                    ProgramSpvPathTemp);
+        if (err != CL_SUCCESS)
+          return err;
+
+        pocl_cache_create_program_cachedir(program, device_i, Binary,
+                                           BinarySize, ProgramBcPath);
+        convertProgramBcToSpv(ProgramBcPath, ProgramSpvPath);
+        pocl_rename(ProgramSpvPathTemp, ProgramSpvPath);
+
+        POCL_RETURN_ERROR_ON(
+            (readProgramSpv(program, device_i, ProgramSpvPath) != CL_SUCCESS),
+            CL_BUILD_PROGRAM_FAILURE,
+            "Could not read compiled program.spv at %s\n", ProgramSpvPath);
+      } else // not SPIRV and not LLVM IR spir
+#endif
+        POCL_RETURN_ERROR_ON(
+            1, CL_BUILD_PROGRAM_FAILURE,
+            "the binary supplied to level0 driver is not SPIR-V, "
+            "and it's not a recognized binary type\n");
+    }
+  }
+
+  assert(program->program_il != nullptr);
+  assert(program->program_il_size > 0);
+
+  if (link_program) {
+    return Device->createProgram(program, device_i);
+  } else {
+    // only final (linked) programs have  ZE module
+    assert(program->data[device_i] == nullptr);
+    return CL_SUCCESS;
+  }
+}
+
+int pocl_level0_link_program(cl_program program, cl_uint device_i,
+                             cl_uint num_input_programs,
+                             const cl_program *input_programs,
+                             int create_library) {
+#ifdef SPIRV_LINK
```

Comment:
Are these necessary? Don't we disable the whole driver if SPIRV_LINK is not available? Or do you plan to support binaries-only loading with this?

Reply:
> Don't we disable the whole driver if SPIRV_LINK is not available

Currently yes, but it might be useful to make the driver compilable without `spirv-link`. It's only used in `clLinkProgram`, which (AFAIK) is not a heavily used function.

Reply:
removed the ifdef